### ISSUE TITLE
Write a short cookbook entry how to make use of the ConfigCacheFactory in 2.7

### ISSUE THREAD

**Issue by mpdude** [symfony/symfony-docs#5136]
... as soon as symfony/symfony#14178 is merged :)

Mention also:

* Ability to use a service to check for the freshness of cache entries (introduced in Symfony 2.8, see https://github.com/symfony/symfony/pull/15738)
* `ClassExistenceResource` (introduced in Symfony 3.2, see https://github.com/symfony/symfony/pull/20121)

**Comment by xabbuh** [comment]
symfony/symfony#14178 is merged. This can be taken. :)

**Comment by mpdude** [comment]
I know :-)

Want to have example code for it first.

**Comment by WouterJ** [comment]
Instead of a cookbook entry, what about adding a new entry to `components/config`?

**Comment by mpdude** [comment]
https://github.com/symfony/symfony/pull/15692 finally makes the ConfigCacheFactory easily available and useful via a new tag.

**Comment by javiereguiluz** [comment]
Just asking: is this something we want to document ... or is this something rarely used or too internal?

If someone understands this feature, please contribute a pull request or explain it to me with some real-world example and I'll add the docs. Thank you!

**Comment by mpdude** [comment]
Oh, it's a shame this has been on my to-do list since 2015 and I did not find the time to work on it 😞 

Of course, this is an rather low level and internal feature. As with many other of those, they are what makes Symfony such a powerful framework, as you have a lot of options for extending and adapting it. On the other hand, most of these features are sparingly documented or have no documentation at all, besides the API docblocks. This makes it difficult for intermediate users to leave the beaten paths, as you need to develop a more thorough understanding how parts of the framework work internally. 

Having said that, I'll try to give you a brief outline of what this feature is about, so we can better discuss the issue.

The `ConfigCache` is used internally by Symfony whenever something "expensive" is generated from config files or other resources and can be kept around for later use.

Off my head, the most relevant examples are the router, translator and service container. All three read various files (YAML, XML, translation catalogues, ...) and end up dumping generated PHP code. It's the `ConfigCache` that dumps this PHP code to some file (so it can be `included` in later requests).

Along with these cache files, the `ConfigCache` keeps a list of "resources" that were used to generate whatever is in the cache itself. Before the symfony/symfony#14178 change, these resources could (basically) only be files that were recorded along with a timestamp of their last modification. When in `kernel.debug` mode, the `ConfigCache` will upon every request work this list of "resources" and check each single one whether or not it has changed. That way, the router or translator can automatically be refreshed when a file like `routing.yml` or a message cataloge is updated.

With the above change, the `ConfigCache` is no longer required to be able to perform this check by itself – it can delegate the check to services from the DIC. This makes it possible to create your own loaders to, for example, load routes or, more importantly, translations from a database.

I contributed this feature because back then I was working on a project where translations for more than 20 languages had to be kept in a database (supporting a collaborative workflow to keep a bigger application i18nized). That change made it possible to load the translation from the DB and even update the message catalogues at runtime when translations were updated. As there were several open issues over at symfony/symfony at that time that were requesting this possibility, I considered it useful for others as well.

This feature may also be useful if you're using the `ConfigCache` to cache something else in your own project/application where "resources" are not only plain files. IIRC there was also some caching issue in symfony/AsseticBundle that I though could be solved with it. But to be honest, I lost sight of that one...

Does that somehow help, @javiereguiluz?

**Comment by javiereguiluz** [comment]
@mpdude thank you so much for these detailed insights!! I'm still not able to add these docs, but hopefully someone else will. Thanks!

**Comment by javiereguiluz** [comment]
Mention `ClassExistenceResource` too (introduced in Symfony 3.2, see https://github.com/symfony/symfony/pull/20121)

**Comment by javiereguiluz** [comment]
Sadly we must close this issue as "won't fix". We couldn't add these docs on time and now it's too late, because the focus of Symfony Docs has moved completely out of this kind of contents mostly related to stand-alone components. Sorry!

**Comment by wouterj** [comment]
I think this is a very nice feature to document, also as framework user. It has a clear use-case:

> I contributed this feature because back then I was working on a project where translations for more than 20 languages had to be kept in a database (supporting a collaborative workflow to keep a bigger application i18nized). That change made it possible to load the translation from the DB and even update the message catalogues at runtime when translations were updated.

However, it seems very advanced. So (a) we need a little sample application to play with this feature before we can document and (b) we should find a place to document something more advanced like this..

**Comment by faizanakram99** [comment]
Legit use case which was solved by this, https://github.com/symfony/symfony/issues/45991#issuecomment-1171411228
